galahq/gala
# Use docker for development environment

### Context

The objective is to containerize the app. Making it easier to develop locally and provide a consistent environment between local and prod. 

The scope of this PR is to add a docker development environment to the app. Prod will not run docker containers, yet.

### Notes

- README is updated with docker instructions and helpful commands.

#### Postgres and Redis
- Removed Postgres and Redis as dependencies. These are now part of the docker compose network. Both their ports 
 are exposed to your machine so you can connect to them directly from outside the docker network. e.g. `psql -h localhost -p 5432 -U gala` connects to the Postgres database that's running inside the docker network.

- `db/structure.sql` is ran automatically on `docker compose up` when the database volume is empty. This means we no longer need to run `rails db:setup` manually. `rails db:seed` now runs automatically via `entrypoint.sh`.

#### Why have Ruby and Node dependencies when there's docker?

- Left Ruby 2.7.6 and Node 12.5 as dependencies to allow you to run the feature tests successfully. I was unable to make the feature tests pass when running inside the docker network. Selenium seems to hang at a certain point, which makes them fail. Because of this, all tests should be ran outside of the docker network, for now. The `selenium` service in the docker compose file is commented out for this reason.

- Keeping Ruby as a dependency also means will still be able to use Rails tasks without being inside a container. E.g. using `bundle exec rails g migration AddNewColumn` to scaffold a migration file.

- Modified the `rails_helper.rb` to fix issues with the `webdrivers` gem. Turns out local, docker selenium, and CI all require different versions of chrome in each environment in order to run.


### Test
1. ensure you have docker downloaded and installed on your machine
2. `git fetch -a && git checkout docker-development-env` to checkout this branch
3. `docker compose up` to run the app
4. navigate to `http://localhost:3000` and interact with the app
5. ensure the browser refreshes automatically on code changes
